app/models/concerns/labware_creators/donor_pooling_calculator.rb

```diff
@@ -130,34 +130,110 @@ def unique_donor_ids(group)
+  end
+
+  def reassign_to_next_pool(args, pool_index, number_of_pools)
+    args[:pool_index] = (pool_index + 1) % number_of_pools
+    assign_well_to_pool(args)
+  end
+
+  def add_to_pool(donor_id, used_donor_ids, pool_index, pools, well)
+    used_donor_ids[pool_index] << donor_id
+    pools[pool_index] << well
+  end
+
+  # Allocates wells to pools. The wells will have grouped by study and project, and now
+  # they will be grouped by unique donor_ids. The wells will be distributed sequentially
+  # to the pools, ensuring that each pool has between 5 and 25 wells.
   #
-  # the result will be:
-  # [[3], [1], [2], [4, 5], [6, 7], [8, 9]]
+  # If the number of wells is 96 and the number of pools is 8, then
+  # each pool will have 12 wells
+  # [[12], [12], [12], [12], [12], [12], [12], [12]]
   #
-  # for which the steps are:
-  # [[1, 2, 3], [4, 5], [6, 7, 8, 9]] -> 3 pools (input)
-  # [[4, 5], [6, 7], [8, 9], [1, 2, 3]] -> 4 pools
-  # [[3], [4, 5], [6, 7], [8, 9], [1, 2]] -> 5 pools
-  # [[3], [1], [2], [4, 5], [6, 7], [8, 9]] -> 6 pools (output)
   #
-  # @param groups [Array<Array<Well>>] Array of well groups to be distributed.
-  # @return [Array<Array<Well>>] Array of distributed groups.
-  def distribute_groups_across_pools(groups, number_of_pools)
-    groups = groups.dup
-    groups.sort_by!(&:size)
-    while groups.any? && groups.last.size > 1 && groups.size < number_of_pools
-      largest = groups.pop # last
-      splits = largest.each_slice((largest.size / 2.0).ceil).to_a
-      groups.concat(splits).sort_by!(&:size)
+  # If the number of wells is 96 and the number of pools is 7, then
+  # the first 5 pools will have 14 wells and the last 2 pools will have 13 wells
+  # [[14], [14], [14], [14], [14], [13], [13]]
+  #
+  # If the number of wells is 24 and the number of pools is 5, then
+  # an error will be raised because each pool must have at least 5 wells
+  #
+  # @param wells [Array<Well>] The wells to be allocated to pools.
+  # @param number_of_pools [Integer] The number of pools to distribute the wells into.
+  # @return [Array<Array<Well>>] An array of pools, between 1 and 8, each containing between 5 and 25 wells.
+  #
+  def allocate_wells_to_pools(wells, number_of_pools)
+    pools = Array.new(number_of_pools) { [] }
+    used_donor_ids = Array.new(number_of_pools) { [] }
+
+    validate_number_of_pools(wells, number_of_pools)
+    depth = 0
+
+    # Assign wells to pools
+    wells.each_with_index do |well, index|
+      # Start assigning wells starting from the pool corresponding to the well's index
+      assign_well_to_pool(
+        {
+          well: well,
+          pools: pools,
+          used_donor_ids: used_donor_ids,
+          pool_index: index % number_of_pools,
+          number_of_pools: number_of_pools,
+          depth: depth
+        }
+      )
+    end
+
+    if pools.any? { |pool| pool.size < 5 || pool.size > 25 }
```

Comment:
use constants for the 5 and 25 values e.g. MIN_POOL_SIZE
(and in the error msg)